flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/math/distributed/DistributedRowMatrix.scala

```diff
@@ -0,0 +1,167 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.math.distributed
+
+import java.lang
+
+import breeze.linalg.{CSCMatrix => BreezeSparseMatrix, Matrix => BreezeMatrix, Vector => BreezeVector}
+import org.apache.flink.api.common.functions.RichGroupReduceFunction
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.scala._
+import org.apache.flink.ml.math.{Matrix => FlinkMatrix, _}
+import org.apache.flink.util.Collector
+import org.apache.flink.ml.math.Breeze._
+import scala.collection.JavaConversions._
+
+/**
+  *
+  * @param numRowsOpt If None, will be calculated from the DataSet.
+  * @param numColsOpt If None, will be calculated from the DataSet.
+  */
+class DistributedRowMatrix(data: DataSet[IndexedRow],
+                           numRowsOpt: Option[Int] = None,
+                           numColsOpt: Option[Int] = None)
+    extends DistributedMatrix {
+
+  lazy val getNumRows: Int = numRowsOpt match {
+    case Some(rows) => rows
+    case None => data.count().toInt
+  }
+
+  lazy val getNumCols: Int = numColsOpt match {
+    case Some(cols) => cols
+    case None => calcCols
+  }
+
+  val getRowData = data
+
+  private def calcCols: Int =
+    data.first(1).collect().headOption match {
+      case Some(vector) => vector.values.size
+      case None => 0
+    }
+
+  /**
+    * Collects the data in the form of a sequence of coordinates associated with their values.
+    * @return
+    */
+  def toCOO: Seq[(Int, Int, Double)] = {
+
+    val localRows = data.collect()
+
+    for (IndexedRow(rowIndex, vector) <- localRows;
+    (columnIndex, value) <- vector) yield (rowIndex, columnIndex, value)
+  }
+
+  /**
+    * Collects the data in the form of a SparseMatrix
+    * @return
+    */
+  def toLocalSparseMatrix: SparseMatrix = {
+    val localMatrix = SparseMatrix.fromCOO(this.getNumRows, this.getNumCols, this.toCOO)
+    require(localMatrix.numRows == this.getNumRows)
+    require(localMatrix.numCols == this.getNumCols)
+    localMatrix
+  }
+
+  //TODO: convert to dense representation on the distributed matrix and collect it afterward
+  def toLocalDenseMatrix: DenseMatrix = this.toLocalSparseMatrix.toDenseMatrix
+
+  def byRowOperation(
+      fun: (Vector, Vector) => Vector, other: DistributedRowMatrix): DistributedRowMatrix = {
+    val otherData = other.getRowData
+    require(this.getNumCols == other.getNumCols)
+    require(this.getNumRows == other.getNumRows)
+
+    val ev1: TypeInformation[Int] = TypeInformation.of(classOf[Int])
+
+    val result = this.data
+      .fullOuterJoin(otherData)
+      .where(_.rowIndex)
+      .equalTo(_.rowIndex)(ev1)(
```

Comment:
Good to know. I will have to fix this in the other PR too because I used it a lot. Thanks.